tests/tests/swan_tests.cpp

```diff
@@ -0,0 +1,274 @@
+
+#include <fc/crypto/digest.hpp>
+
+#include "../common/database_fixture.hpp"
+
+using namespace graphene::chain;
+using namespace graphene::chain::test;
+
+BOOST_FIXTURE_TEST_SUITE( swan_tests, database_fixture )
+
+/**
+ *  This test sets up the minimum condition for a black swan to occur but does
+ *  not test the full range of cases that may be possible during a black swan.
+ */
+BOOST_AUTO_TEST_CASE( black_swan )
+{ try {
+      ACTORS((borrower)(borrower2)(feedproducer));
+
+      const auto& bitusd = create_bitasset("USDBIT", feedproducer_id);
+      const asset_id_type bitusd_id = bitusd.id;
+      const auto& core   = asset_id_type()(db);
+
+      int64_t init_balance(1000000);
+
+      transfer(committee_account, borrower_id, asset(init_balance));
+      transfer(committee_account, borrower2_id, asset(init_balance));
+      update_feed_producers(bitusd, {feedproducer.id});
+
+      price_feed current_feed;
+      current_feed.settlement_price = bitusd.amount(100) / core.amount(100);
+
+      // starting out with price 1:1
+      publish_feed(bitusd, feedproducer, current_feed);
+
+      // start out with 2:1 collateral
+      borrow(borrower, bitusd.amount(1000), asset(2000));
+      borrow(borrower2, bitusd.amount(1000), asset(4000));
+
+      BOOST_REQUIRE_EQUAL( get_balance(borrower, bitusd), 1000 );
+      BOOST_REQUIRE_EQUAL( get_balance(borrower2, bitusd), 1000 );
+      BOOST_REQUIRE_EQUAL( get_balance(borrower , core), init_balance - 2000 );
+      BOOST_REQUIRE_EQUAL( get_balance(borrower2, core), init_balance - 4000 );
+
+      current_feed.settlement_price = bitusd.amount( 100 ) / core.amount(200);
+      publish_feed( bitusd, feedproducer, current_feed );
+
+      /// this sell order is designed to trigger a black swan
+      create_sell_order( borrower2, bitusd.amount(1000), core.amount(3000) );
+
+      BOOST_CHECK( bitusd.bitasset_data(db).has_settlement() );
+
+      force_settle(borrower, bitusd.amount(100));
+
+      // make sure pricefeeds expire
+      generate_blocks(db.head_block_time() + GRAPHENE_DEFAULT_PRICE_FEED_LIFETIME);
+      generate_blocks( HARDFORK_CORE_216_TIME );
+      generate_blocks(2);
+
+      FC_ASSERT( bitusd_id(db).bitasset_data(db).current_feed.settlement_price.is_null() );
+      force_settle(borrower_id(db), asset(100, bitusd_id));
+
+      current_feed.settlement_price = bitusd_id(db).amount(100) / asset_id_type()(db).amount(150);
+      publish_feed(bitusd_id(db), feedproducer_id(db), current_feed);
+
+      BOOST_TEST_MESSAGE( "Verify that we cannot borrow after black swan" );
+      GRAPHENE_REQUIRE_THROW( borrow(borrower_id(db), asset(1000, bitusd_id), asset(2000)), fc::exception )
+      trx.operations.clear();
+} catch( const fc::exception& e) {
+      edump((e.to_detail_string()));
+      throw;
+   }
+}
+
+/**
+ * Black swan occurs when price feed falls, triggered by settlement
+ * order.
+ */
+BOOST_AUTO_TEST_CASE( black_swan_issue_346 )
```

Comment:
unsure about the reference to issue 346 here, it seems to not be related in bitshares-core or cryptonomex repos.

Reply:
This was moved here from the original https://github.com/bitshares/bitshares-core/blob/master/tests/tests/operation_tests.cpp#L287 , I didn't check what the 346 refers to. It seems that it indeed references the cryptonomes repo - https://github.com/cryptonomex/graphene/commit/9d5a5dd21ccfab34ef425c418718cac4bc02fe2b

Reply:
you are right it actually came from here https://github.com/cryptonomex/graphene/issues/346, at first saw i thought it was unrelated but going lower into the issue bytemaster explains. thanks.